diff -rupN bfd/bfd-in2.h bfd/bfd-in2.h
--- bfd/bfd-in2.h	2012-08-18 22:45:19.159911234 +0200
+++ bfd/bfd-in2.h	2012-08-18 22:29:24.000000000 +0200
@@ -4056,6 +4056,10 @@ instructions  */
 instructions  */
   BFD_RELOC_AVR_6_ADIW,
 
+/* This is a 7 bit reloc for the AVR that stores offset for 16bit sts/lds
+instructions meant for tiny parts  */
+  BFD_RELOC_AVR_7_LDS16,
+
 /* Renesas RX Relocations.  */
   BFD_RELOC_RX_NEG8,
   BFD_RELOC_RX_NEG16,
diff -rupN bfd/libbfd.h bfd/libbfd.h
--- bfd/libbfd.h	2011-08-17 02:39:39.000000000 +0200
+++ bfd/libbfd.h	2012-08-18 22:29:22.000000000 +0200
@@ -1834,6 +1834,7 @@ static const char *const bfd_reloc_code_
   "BFD_RELOC_AVR_LDI",
   "BFD_RELOC_AVR_6",
   "BFD_RELOC_AVR_6_ADIW",
+  "BFD_RELOC_AVR_7_LDS16",
   "BFD_RELOC_RX_NEG8",
   "BFD_RELOC_RX_NEG16",
   "BFD_RELOC_RX_NEG24",
